### PR TITLE
Add Spatial Transformation Attack

### PR DESCRIPTION
**Pull request by yenchenlin**
This pull request implements the spatial transformation-based attack proposed in [1], addressing #490 . 

Here is the procedure of the attack:
1. randomly sample `n_samples` set of different translation & rotation parameters
2. apply the transformations to the testing set and choose the worst of `n_samples` transformation as the attack

A notebook demo of the attack can be found [here](http://nbviewer.jupyter.org/github/yenchenlin/cleverhans/blob/transformation-attack/cleverhans_tutorials/mnist_tutorial_tran.ipynb), where the testing accuracy of the classifier drops from **98.74%** to **59.61%** (Results in [1]: **99.31%** to **73.32%**)

[1] A Rotation and a Translation Suffice: Fooling CNNs with Simple Transformations, Engstorm et al.

### CODE REVIEW

**Comment by catherio** [comment]
You've got a few PEP8 errors due to lines being too long, let me see if I can clean those up for you.

**Comment by catherio** [comment]
@nottombrown Would you mind reviewing this to compare it with your implementation?

**Comment by yenchenlin** [comment]
@catherio Thanks a lot!

**Comment by nottombrown** [comment]
Sorry for the slow response. I'll take a look at this over the weekend.

**Comment by yenchenlin** [comment]
@nottombrown Thanks!

**Comment by nottombrown** [comment]
Sorry for the delay. Scheduled some time on Friday to look at this.

**Comment by yenchenlin** [comment]
@nottombrown No worry take your time!

File: cleverhans/attacks.py [hunk 1]
```diff
@@ -1821,3 +1821,109 @@ def loss_fn(x, label):
...
+        self.n_dxs = n_dxs
+        self.dy_min = dy_min
+        self.dy_max = dy_max
+        self.n_dys = n_dys
+        self.angle_min = angle_min
+        self.angle_max = angle_max
+        self.n_angles = n_angles
+
+        if self.angle_min < -30 or self.angle_max > 30:
+            raise ValueError("The value of angle must be bounded "
```
Comment (nottombrown) [inline cleverhans/attacks.py]: Why do we need to enforce this bound?

File: cleverhans/attacks_tf.py [hunk 2]
```diff
@@ -1918,3 +1918,223 @@ def margin_logit_loss(model_logits, label, num_classes=10):
         logits_with_target_label_neg_inf, axis=-1)
     loss = highest_nonlabel_logits - label_logits
     return loss
+
+
+def _transformer(U, theta, out_size, name='SpatialTransformer', **kwargs):
```
Comment (nottombrown) [inline cleverhans/attacks_tf.py]: I have trouble following what's going on in the `_transformer` method.

Is it doing something similar to the transform method from the original Engstrom et al. paper?
([our slightly modified version](https://github.com/google/unrestricted-adversarial-examples/blob/1e38f25ec4aee5d5105580f3501335e792a837dd/unrestricted-advex/unrestricted_advex/attacks.py#L240) copied below) 
```
def apply_transformation(x, transform, image_height, image_width):
  # Map a transformation onto the input
  trans_x, trans_y, rot = tf.unstack(transform, axis=1)
  rot *= np.pi / 180  # convert degrees to radians

  # Pad the image to prevent two-step rotation / translation
  # resulting in a cropped image
  x = tf.pad(x, [[0, 0], [50, 50], [50, 50], [0, 0]], 'CONSTANT')

  # rotate and translate image
  ones = tf.ones(shape=tf.shape(trans_x))
  zeros = tf.zeros(shape=tf.shape(trans_x))
  trans = tf.stack([ones, zeros, -trans_x,
                    zeros, ones, -trans_y,
                    zeros, zeros], axis=1)
  x = tf.contrib.image.rotate(x, rot, interpolation='BILINEAR')
  x = tf.contrib.image.transform(x, trans, interpolation='BILINEAR')
  return tf.image.resize_image_with_crop_or_pad(
    x, image_height, image_width)
```

Comment (nottombrown) [reply cleverhans/attacks_tf.py]: If so, maybe use the `tf.contrib.image` library?

If not, perhaps you could add tests for the `_transformer` method to show the usage and correctness?

File: cleverhans/attacks_tf.py [hunk 3]
```diff
@@ -1918,3 +1918,223 @@ def margin_logit_loss(model_logits, label, num_classes=10):
...
+                input_transformed,
+                tf.stack([num_batch, out_height, out_width, num_channels]))
+            return output
+
+    with tf.variable_scope(name):
+        output = _transform(theta, U, out_size)
+        return output
+
+
+def spm(x, model, batch_size=128, y=None, n_samples=10, dx_min=-0.1,
```
Comment (nottombrown) [inline cleverhans/attacks_tf.py]: I expect that most people will want to use the full grid (because it is the strongest attack). Perhaps default to the full grid if `n_samples` is `None`.

If you don't think that is the best behavior, we should at least document that the caller should set `n_samples = n_dxs * n_dys * n_angles` to get the full grid.

Comment (nottombrown) [reply cleverhans/attacks_tf.py]: Otherwise I find this method clear and easy to read.

File: cleverhans/attacks.py [hunk 4]
```diff
@@ -1821,3 +1821,109 @@ def loss_fn(x, label):
...
+                      along x-axis.
+        :param dy_min: (optional float) Minimum translation ratio along y-axis.
+        :param dy_max: (optional float) Maximum translation ratio along y-axis.
+        :param n_dys: (optional int) Number of discretized translation ratios
+                      along y-axis.
+        :param angle_min: (optional float) Largest counter-clockwise rotation
+                          angle.
+        :param angle_max: (optional float) Largest clockwise rotation angle.
+        :param n_angles: (optional int) Number of discretized angles.
+        """
```
Comment (nottombrown) [inline cleverhans/attacks.py]: I like this documentation and parameterization of the arguments. 👍

File: cleverhans/attacks.py (hunk 1 above)
Comment (yenchenlin) [reply cleverhans/attacks.py]: We don't need to have this, I added it just because this is the bound in the paper.

**Comment by nottombrown** [comment]
Let me know if you would like any clarification @yenchenlin

**Comment by yenchenlin** [comment]
Hello @nottombrown , thanks for your helpful review! I've simplified the code (a lot) based on your suggestions. 

I have two questions now:
1. Is the padding part of your example tranformation function essential? I tried it but 
```
return tf.image.resize_image_with_crop_or_pad(x, image_height, image_width)
```
keep outputting an error saying that Dimension is -100 but it should > 0. I guess -100 comes from 28 (original size) - 128 (padded size) = -100. 

2. Can you suggest me what kind of test cases should I add? Seems like different attacks have different unit tests.

**Comment by goodfeli** [comment]
hi @yenchenlin , will you be able to finish this PR?

**Comment by yenchenlin** [comment]
@goodfeli yes, working on it. Will update the code later today.

**Comment by yenchenlin** [comment]
Hello @nottombrown and @goodfeli , just added the tests and the padding before transformation to prevent corner cutoffs.

File: cleverhans/attacks_tf.py [hunk 5]
```diff
@@ -1918,3 +1918,69 @@ def margin_logit_loss(model_logits, label, num_classes=10):
...
+    angle *= np.pi / 180
+    height = x.get_shape().as_list()[1]
+    width = x.get_shape().as_list()[2]
+    M = np.array([1, 0, -dx*height,
+                  0, 1, -dy*width, 0, 0] * batch_size, dtype=np.float32)
+    theta = tf.constant(M, shape=(batch_size, 8))
+
+
+    # Pad the image to prevent two-step rotation / translation
+    x = tf.pad(x, [[0, 0], [50, 50], [50, 50], [0, 0]], 'CONSTANT')
```
Comment (nottombrown) [inline cleverhans/attacks_tf.py]: It could make sense to have the padding depend on the image size. If the image is large, we could still end up with chopped corners.

**Comment by yenchenlin** [comment]
Hello @nottombrown , just sampled 100 adversarial images from MNIST:
download

I just rebased the branch to master but not sure whether this is what you mean ... sorry for making the commit logs on github looks a little bit weird.

**Comment by nottombrown** [comment]
Ok, travis seems to be failing now, but lgtm pending travis passing

**Comment by nottombrown** [comment]
@yenchenlin, I have a commit that seems to solve the issue here:
https://github.com/tensorflow/cleverhans/commit/04eec9ff4ad87df669d62c6952e8a2bb470f715c

Want to pull my change into this PR and then I can merge it into master?

**Comment by yenchenlin** [comment]
@nottombrown Oh yeah thanks for the catch! Would be happy to do so.
